File: src/main/scala/arimitsu/sf/cql/v3/util/Notation.scala
package arimitsu.sf.cql.v3.util

import java.nio.ByteBuffer
import arimitsu.sf.cql.v3.Consistency
import java.net.InetSocketAddress
import java.util.UUID

/**
 * Created by sxend on 14/06/07.
 */
object Notation {

  case class OptionNotation(id: Short, value: Option[Any])

  def getInt(buffer: ByteBuffer): Int = {
    buffer.getInt
  }

  def getLong(buffer: ByteBuffer): Long = {
    buffer.getLong
  }

  def getShort(buffer: ByteBuffer): Short = {
    buffer.getShort
  }

  def getString(buffer: ByteBuffer): String = {
    getString(buffer, buffer.getShort)
  }

  def getString(buffer: ByteBuffer, length: Int): String = {
    val bytes = new Array[Byte](length)
    buffer.get(bytes)
    new String(bytes)
  }

  def getLongString(buffer: ByteBuffer): String = {
    val length = getInt(buffer)
    getString(buffer, length)
  }

  def getUUID(buffer: ByteBuffer): UUID = {
    val bytes = new Array[Byte](16)
    buffer.get(bytes)
    UUID.nameUUIDFromBytes(bytes)
  }

  def getStringList(buffer: ByteBuffer): List[String] = {
    (0 until buffer.getShort).map {
      _ => getString(buffer)
    }.toList
  }

  def getBytes(buffer: ByteBuffer): Option[Array[Byte]] = {
    val length = getInt(buffer)
    getBytes(buffer, length)
  }

  def getShortBytes(buffer: ByteBuffer): Array[Byte] = {
    val length = getShort(buffer)
    getBytes(buffer, length).get
  }

  def getBytes(buffer: ByteBuffer, length: Int): Option[Array[Byte]] = {
    Option(
      if (length > 0) {
        val bytes = new Array[Byte](length)
        buffer.get(bytes)
        bytes
      } else {
        null
      }
    )
  }

  def getOption(optionParser: ByteBuffer => Option[Any])(buffer: ByteBuffer): OptionNotation = {
    val id = getShort(buffer)
    OptionNotation(id, optionParser(buffer))
  }

  def getOptionList(listParser: ByteBuffer => List[OptionNotation])(buffer: ByteBuffer): List[OptionNotation] = {
    listParser(buffer)
  }

  def getINet(buffer: ByteBuffer): InetSocketAddress = {
    val addrArea = new Array[Byte](buffer.get)
    buffer.get(addrArea)
    InetSocketAddress.createUnresolved(new String(addrArea), getInt(buffer))
  }

  def getConsistency(buffer: ByteBuffer): Consistency = {
    Consistency.valueOf(getShort(buffer))
  }

  def getStringMap(buffer: ByteBuffer): Map[String, String] = {
    (0 until buffer.getShort).map {
      _ => (getString(buffer), getString(buffer))
    }.toMap
  }

  def getStringMultiMap(buffer: ByteBuffer): Map[String, List[String]] = {
    (0 until buffer.getShort).map {
      _ => (getString(buffer), getStringList(buffer))
    }.toMap
  }

  def toStringMap(maps: Map[String, String]): Array[Byte] = {
    Array.concat(short2Bytes(maps.size.toShort),
      maps.map {
        v => Array.concat(toString(v._1), toString(v._2))
      }.foldLeft(Array.empty[Byte]) {
        (a, b) => Array.concat(a, b)
      })
  }

  def toString(str: String): Array[Byte] = {
    val bytes = str.getBytes("UTF-8")
    Array.concat(short2Bytes(bytes.length.toShort), bytes)
  }

  def short2Bytes(s: Short): Array[Byte] = {
    val bytes = new Array[Byte](2)
    bytes.update(0, (0xff & (s >> 8)).toByte)
    bytes.update(1, (0xff & s).toByte)
    bytes
  }
}

File: src/main/scala/arimitsu/sf/cql/v3/util/Notation.java
package arimitsu.sf.cql.v3.util;

import arimitsu.sf.cql.v3.Consistency;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Created by sxend on 14/06/07.
 */
public class Notation {

    public static class OptionNotation<A> {
        public final short id;
        public final A value;

        public OptionNotation(short id, A value) {
            this.id = id;
            this.value = value;
        }
    }

    public static interface OptionParser<A> {
        public A parse(ByteBuffer buffer);
    }

    public static short getShort(ByteBuffer buffer) {
        return buffer.getShort();
    }

    public static String getString(ByteBuffer buffer) {
        return getString(buffer, buffer.getShort());
    }

    public static final String EMPTY = "";

    public static String getString(ByteBuffer buffer, int length) {
        if (length <= 0) return EMPTY;
        byte[] bytes = new byte[length];
        buffer.get(bytes);
        return new String(bytes);
    }

    public static String getLongString(ByteBuffer buffer) {
        return getString(buffer, buffer.getInt());
    }

    public static UUID getUUID(ByteBuffer buffer) {
        byte[] bytes = new byte[16];
        buffer.get(bytes);
        return UUID.nameUUIDFromBytes(bytes);
    }

    private static final Constructor<UUID> uuidConstructor;

    static {
        Constructor<UUID> c = null;
        try {
            c = UUID.class.getDeclaredConstructor(byte[].class);
            c.setAccessible(true);
        } catch (NoSuchMethodException e) {
        }
        uuidConstructor = c;
    }

    public static UUID toUUID(byte[] bytes) {
        try {
            return uuidConstructor.newInstance(bytes);
        } catch (InvocationTargetException | InstantiationException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }

    public static List<String> getStringList(ByteBuffer buffer) {
        List<String> list = new ArrayList<>();
        for (int i = 0, length = buffer.getShort(); i < length; i++) {
            list.add(getString(buffer));
        }
        return list;
    }

    public static List<String> getLongStringList(ByteBuffer buffer) {
        List<String> list = new ArrayList<>();
        for (int i = 0, length = buffer.getInt(); i < length; i++) {
            list.add(getLongString(buffer));
        }
        return list;
    }

    public static byte[] getBytes(ByteBuffer buffer) {
        return getBytes(buffer, buffer.getInt());
    }

    public static byte[] getShortBytes(ByteBuffer buffer) {
        return getBytes(buffer, buffer.getShort());
    }

    public static byte[] getBytes(ByteBuffer buffer, int length) {
        if (length > 0) {
            byte[] bytes = new byte[length];
            buffer.get(bytes);
            return bytes;
        } else {
            return null;
        }
    }

    public static <A> OptionNotation<A> getOption(ByteBuffer buffer, OptionParser<A> parser) {
        short id = getShort(buffer);
        return new OptionNotation<>(id, parser.parse(buffer));
    }

    public static <A> List<OptionNotation<A>> getOptionList(ByteBuffer buffer, OptionParser<A> parser) {
        List<OptionNotation<A>> list = new ArrayList<>();
        for (int i = 0, length = buffer.getShort(); i < length; i++) {
            list.add(getOption(buffer, parser));
        }
        return list;
    }

    public static InetAddress getINet(ByteBuffer buffer) {
        byte[] addrArea = new byte[buffer.get()];
        buffer.get(addrArea);
        return toInet(addrArea);
    }

    public static InetAddress toInet(byte[] bytes) {
        try {
            return InetAddress.getByAddress(bytes);
        } catch (UnknownHostException e) {
            throw new RuntimeException(e);
        }
    }

    public static Consistency getConsistency(ByteBuffer buffer) {
        return Consistency.valueOf(getShort(buffer));
    }

    public static Map<String, String> getStringMap(ByteBuffer buffer) {
        Map<String, String> map = new HashMap<>();
        for (int i = 0, length = buffer.getShort(); i < length; i++) {
            map.put(getString(buffer), getString(buffer));
        }
        return map;
    }

    public static Map<String, List<String>> getStringMultiMap(ByteBuffer buffer) {
        Map<String, List<String>> map = new HashMap<>();
        for (int i = 0, length = buffer.getShort(); i < length; i++) {
            map.put(getString(buffer), getStringList(buffer));
        }
        return map;
    }

    public static byte[] toStringMap(Map<String, String> maps) {
        byte[] mapLength = short2Bytes((short) maps.size());
        byte[] result = new byte[0];
        for (Map.Entry<String, String> entry : maps.entrySet()) {
            result = join(result, join(toString(entry.getKey()), toString(entry.getValue())));
        }
        return join(mapLength, result);
    }

    public static byte[] toLongString(String str) {
        byte[] bytes;
        try {
            bytes = str.getBytes("UTF-8");
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
        return join(int2Bytes(bytes.length), bytes);
    }

    public static byte[] toString(String str) {
        byte[] bytes;
        try {
            bytes = str.getBytes("UTF-8");
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
        int length = bytes.length;
        return join(short2Bytes((short) length), bytes);
    }

    public static byte[] short2Bytes(short s) {
        byte[] bytes = new byte[2];
        bytes[0] = (byte) (0xff & (s >>> 8));
        bytes[1] = (byte) (0xff & s);
        return bytes;
    }

    public static byte[] int2Bytes(int s) {
        byte[] bytes = new byte[4];
        bytes[0] = (byte) (0xff & (s >>> 24));
        bytes[1] = (byte) (0xff & (s >>> 16));
        bytes[2] = (byte) (0xff & (s >>> 8));
        bytes[3] = (byte) (0xff & s);
        return bytes;
    }

    public static byte[] long2Bytes(long s) {
        byte[] bytes = new byte[8];
        bytes[0] = (byte) (0xff & (s >>> 56));
        bytes[1] = (byte) (0xff & (s >>> 48));
        bytes[2] = (byte) (0xff & (s >>> 40));
        bytes[3] = (byte) (0xff & (s >>> 32));
        bytes[4] = (byte) (0xff & (s >>> 24));
        bytes[5] = (byte) (0xff & (s >>> 16));
        bytes[6] = (byte) (0xff & (s >>> 8));
        bytes[7] = (byte) (0xff & s);
        return bytes;
    }

    public static byte[] join(byte[] byte1, byte[] byte2) {
        byte[] resultBytes = new byte[byte1.length + byte2.length];
        System.arraycopy(byte1, 0, resultBytes, 0, byte1.length);
        System.arraycopy(byte2, 0, resultBytes, byte1.length, byte2.length);
        return resultBytes;
    }

    public static long getLong(byte[] bytes) {
        long result = 0;
        for (int i = 0; i < bytes.length; i++) {
            result += ((0xff & bytes[i]) << ((bytes.length - i - 1) * 8));
        }
        return result;
    }
}
